src/D2L.CodeStyle.Analysis/MutabilityInspector.cs

```diff
@@ -0,0 +1,97 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+
+namespace D2L.CodeStyle.Analysis {
+
+	public sealed class MutabilityInspector {
+
+		/// <summary>
+		/// A list of known non-valuetype immutable types
+		/// </summary>
+		private static readonly ImmutableHashSet<string> KnownImmutableTypes = new HashSet<string> {
+			nameof(String)
+		}.ToImmutableHashSet();
+
+		/// <summary>
+		/// Determine if a given type is mutable.
+		/// </summary>
+		/// <param name="type">The type to determine mutability for.</param>
+		/// <returns>Whether the type is mutable.</returns>
+		public bool IsTypeMutable(
+			ITypeSymbol type
+		) {
+			if( type.IsValueType ) {
+				return false;
+			}
+
+			if( type.TypeKind == TypeKind.Array ) {
+				return true;
+			}
+
+			if( KnownImmutableTypes.Contains( type.Name ) ) {
+				return false;
+			}
+
+			foreach( var member in type.GetMembers() ) {
+				if( member is IPropertySymbol ) {
+					var prop = member as IPropertySymbol;
+					if( IsPropertyMutable( prop ) || IsTypeMutable( prop.Type ) ) {

```

Comment:
Fail, I think this logic is wrong...
if a field is not mutable, this will continue, even if the type is mutable.